### PR TITLE
Feature/economy

### LINKED ISSUE #20: Add Tucsonians economy functionality
add functionality to keep track of some sort of server currency that accumulates every day you are active on the server. you can use this currency for certain marketplace goodies. there admin commands to add/remove currency from players in case there is an issue in the system.

Idea credit for this goes to Kristen

### PR DESCRIPTION
**Pull request by EricJ98**

### CODE REVIEW

File: cogs/economy.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+from discord.ext import commands
+from datetime import datetime
+import math
+
+class Economy(commands.Cog):
+    def __init__(self, bot):
+        self.bot = bot
+        self.storage = self.bot.get_cog('Storage')
+        self.voice_logs = self.storage.get_voice_log_cache()
+        self.voice_logs = {}
```
Comment (ogunasekara) [inline cogs/economy.py]: REQ: remove this

File: cogs/economy.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        # Create a list of tuples from the dictionary, where each tuple is (key, value)
+        items = self.voice_logs.items()
+        # Sort the list of tuples based on the 'cactus_points' value in descending order
+        sorted_items = sorted(items, key=lambda x: x[1]['cactus_points'], reverse=True)
+        # Return the sorted list of tuples
+        return sorted_items
+
+    # Prints out the point leaderboard 
+    @commands.command()
+    async def leaderboard(self, ctx):
```
Comment (ogunasekara) [inline cogs/economy.py:53]: REQ: add triple quote description for this command

File: cogs/economy.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        sorted_users = self.get_top_ten()
+        for i in range(10):
+            embed.add_field(name=sorted_users[i]['member_obj'].name, value= "", inline=True)
+            embed.add_field(name=sorted_users[i]['cactus_points'], value="", inline=True)
+            embed.add_field(name=sorted_users[i]['total_time'], value="", inline=True)
+        await ctx.send(file=file, embed=embed)
+
+    # Prints out individual stats
+    @commands.command()
+    async def stats(self, ctx):
```
Comment (ogunasekara) [inline cogs/economy.py:72]: REQ: add triple quote description for this command

File: cogs/economy.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        self.storage = self.bot.get_cog('Storage')
+        self.voice_logs = self.storage.get_voice_log_cache()
+        self.voice_logs = {}
+        
+    # Listener that updates when a user joins and leaves a voice channel
+    @commands.Cog.listener()
+    async def on_voice_state_update(self, member, before, after):
+        # First time a user ever joins a voice channel in the server
+        if member.id not in self.voice_logs:
+            self.voice_logs[member.id] = {'member_obj': member,'start_time': datetime.utcnow(), 'total_time': 0, 'cactus_points': 0}
```
Comment (ogunasekara) [inline cogs/economy.py:18]: Q: this also gets called when a member leaves a channel for the first time, should this be only done when someone enters a channel for the first time?

File: cogs/economy.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
+import discord
+import time
+from discord.ext import commands
+from datetime import datetime
+import math
```
Comment (ogunasekara) [inline cogs/economy.py]: NIT: put the math import with the other base imports and add a space before the individual imports

**Review by IPlayRunescape — APPROVED** [review]
LGTM

**Review by ogunasekara — APPROVED** [review]

**Pull request merged**